sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
   }
 
   /**
+   * Statically evaluate an expression containing one or more aggregates on an empty input.
+   */
+  private def evalOnZeroTups(expr : Expression) : Option[Any] = {
+    // AggregateExpressions are Unevaluable, so we need to replace all aggregates
+    // in the expression with the value they would return for zero input tuples.
+    val rewrittenExpr = expr transform {
+      case a @ AggregateExpression(aggFunc, _, _, resultId) =>
+        val resultLit = aggFunc.defaultResult match {
+          case Some(lit) => lit
+          case None => Literal.default(NullType)
+        }
+        Alias(resultLit, "aggVal") (exprId = resultId)
+    }
+    Option(rewrittenExpr.eval())
+  }
+
+  /**
    * Construct a new child plan by left joining the given subqueries to a base plan.
    */
   private def constructLeftJoins(
       child: LogicalPlan,
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
+        val aggOutputExpr = query.asInstanceOf[Aggregate].aggregateExpressions.head
```

Comment:
For example, the (somewhat contrived) query:
```scala
range(0, 10).registerTempTable("a")
val plan = sql("select * from a where (select count(*) x from a aa where aa.id = a.id and aa.id % 2 = 0 having x > 0) > 0")
plan.explain(true)
```
results in the following analyzed plan:
```
== Analyzed Logical Plan ==
id: bigint
Project [id#0L]
+- Filter (scalar-subquery#27 [(id#0L#41L = id#0L)] > cast(0 as bigint))
   :  +- SubqueryAlias scalar-subquery#27 [(id#0L#41L = id#0L)]
   :     +- Project [x#26L,id#0L AS id#0L#41L]
   :        +- Filter (x#26L > cast(0 as bigint))
   :           +- Aggregate [id#0L], [(count(1),mode=Complete,isDistinct=false) AS x#26L,id#0L]
   :              +- Filter ((id#0L % cast(2 as bigint)) = cast(0 as bigint))
   :                 +- SubqueryAlias aa
   :                    +- SubqueryAlias a
   :                       +- Range 0, 10, 1, 8, [id#0L]
   +- SubqueryAlias a
      +- Range 0, 10, 1, 8, [id#0L]
```